src/wire/igmpv2.rs

```diff
@@ -0,0 +1,403 @@
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the message type field.
+    #[inline]
+    pub fn msg_type(&self) -> Message {
+        let data = self.buffer.as_ref();
+        Message::from(data[field::TYPE])
+    }
+
+    /// Return the Max reponse time
+    #[inline]
+    pub fn max_resp_time(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::MAX_RESP_TIME]
+    }
+
+    /// Return the checksum field.
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::CHECKSUM])
+    }
+
+    /// Return the source address field.
+    #[inline]
+    pub fn group_addr(&self) -> Ipv4Address {
+        let data = self.buffer.as_ref();
+        Ipv4Address::from_bytes(&data[field::GROUP_ADDRESS])
+    }
+
+    /// Validate the header checksum.
+    ///
+    /// # Fuzzing
+    /// This function always returns `true` when fuzzing.
+    pub fn verify_checksum(&self) -> bool {
+        if cfg!(fuzzing) {
+            return true;
+        }
+
+        let data = self.buffer.as_ref();
+        checksum::data(data) == !0
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    /// Set the message type field.
+    #[inline]
+    pub fn set_msg_type(&mut self, value: Message) {
+        let data = self.buffer.as_mut();
+        data[field::TYPE] = value.into()
+    }
+
+    /// Set the Max Resp Time field.
+    #[inline]
+    pub fn set_max_resp_time(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::MAX_RESP_TIME] = value
+    }
+
+    /// Set the checksum field.
+    #[inline]
+    pub fn set_checksum(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        NetworkEndian::write_u16(&mut data[field::CHECKSUM], value)
+    }
+
+    /// Set the group address field
+    #[inline]
+    pub fn set_group_address(&mut self, addr: Ipv4Address) {
+        let data = self.buffer.as_mut();
+        let addr_bytes = addr.as_bytes();
+        // TODO: check for host endiannes?
+        let addr_u32: u32 = (addr_bytes[3] as u32) | (addr_bytes[2] as u32) << 8 |
+                            (addr_bytes[1] as u32) << 16 |
+                            (addr_bytes[0] as u32) << 24;
+        NetworkEndian::write_u32(&mut data[field::GROUP_ADDRESS], addr_u32)
+    }
+
+    /// Compute and fill in the header checksum.
+    pub fn fill_checksum(&mut self) {
+        self.set_checksum(0);
+        let checksum = {
+            let data = self.buffer.as_ref();
+            !checksum::data(data)
+        };
+        self.set_checksum(checksum)
+    }
+}
+
+
+/// A high-level representation of an Internet Group Management Protocol v2 header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Repr {
+    MembershipQuery {
+        max_resp_time: u8,
+        group_addr: Ipv4Address,
+    },
+    MembershipReport {
+        group_addr: Ipv4Address,
+        version: ReportVersion,
```

Comment:
Is there any reason to care about report version? Will you ever emit V1 report packets? If no and no, you should drop this field.